### ISSUE TITLE
SSL verify option being ignored

### ISSUE THREAD

**Issue by niedbalski** [gophercloud/gophercloud#1726]
os-client-config supports the verify option (not part of the auth options) [0] to disallow
verification of SSL certs (for the case of self-signed).

If this option is set, then the transport for the default http client should be set as follows:

		tlsConfig := &tls.Config{InsecureSkipVerify: true}
		transport := &http.Transport{TLSClientConfig: tlsConfig}
		client.HTTPClient.Transport = transport

Currently https://github.com/gophercloud/utils/blob/3b35bcb43167f6d8080fa824c480f04b60311b51/openstack/clientconfig/results.go#L31  parses the option, but isn't being applied afterwards for authentication.

[0] https://docs.openstack.org/os-client-config/latest/user/configuration.html#ssl-settings

**Comment by jtopjian** [comment]
@niedbalski Gophercloud doesn't handle the actual implementation of SSL verification. This is left to the transport configuration, as you've done in https://github.com/openstack-exporter/openstack-exporter/pull/33/files.

See here for a more advanced example of configuration: https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/config.go#L90-L222

Please let me know if this helps or if you have any questions.

**Comment by niedbalski** [comment]
@jtopjian Thanks for your input.

The problem is that the authentication (v2/v3 towards keystone) happens at the
time of creating the service client (NewServiceClient -> AuthenticatedClient -> Authenticate(), and there is no a way (neither via a clientoption or auth option) to be explicit on how to set the transport options for SSL and the service client doesn't honors verify == false by default.

In my case, I am creating the service client instance as follows:

	client, err := clientconfig.NewServiceClient("network", &opts)
	if err != nil {
		return nil, err
	}

And since SSL verification is enabled, this fails.

**Comment by jtopjian** [comment]
@niedbalski Ah, I see. It looks like the `NewServiceClient` convenience function might need some work, if that's even possible.

To get around this, you'll have to a more manual client creation: https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/config.go#L144-L152

This will create what's called a Provider Client: a client to interact with the OpenStack cloud (namely, Keystone and the catalog).

Once you have that, you can do your SSL and other transport and http client customizations: https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/config.go#L154-L208

Finally, you can create a Service Client (a client for the individual OpenStack service) using the Provider Client: https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/config.go#L341-L344

It's only when you create the Service Client that the Provider Client will start making HTTP requests, allowing you to configure SSL at any point before that.

Let me know if this helps.

**Comment by niedbalski** [comment]
@jtopjian Hey Joe, yes, indeed it helps. I am inclined to think that from a usability
point of view, it would be expected that if the verify option is set false (which is valid and accepted), it should have an effect over the default http client transport method used by the library.

**Comment by jtopjian** [comment]
Yes, I can see that.

In this case the `Verify` field is meant as a representation of the `verify` field of the yaml file - the implementation is left to the application developer.

Supporting the ability to manage SSL options within Gophercloud is way out of scope of the project. I think the best that can be done here is provide better documentation.

**Comment by niedbalski** [comment]
@jtopjian 

* Do you think it makes sense to fix the utility function NewServiceClient?

**Comment by jtopjian** [comment]
Historically, `NewServiceClient` was something I created on a whim. I figured it could be used for internal or "simple" cases and anything beyond that would require manual client creation.

If it's possible to modify it to enable customization of the provider client, I'm all for fixing it (or merging a PR from someone else). But one concern I have is that it would become something complicated and hard to maintain.
